summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/AsyncBase.scala

```diff
@@ -74,92 +35,21 @@ abstract class AsyncBase[I, O, S, D, RC](maxWaitingFutures: MaxWaitingFutures, m
   def apply(state: S, in: I): Future[TraversableOnce[(Seq[S], Future[TraversableOnce[O]])]]
   def tick: Future[TraversableOnce[(Seq[S], Future[TraversableOnce[O]])]] = Future.value(Nil)
 
-  private lazy val outstandingFutures = Queue.linkedNonBlocking[Future[Unit]]
-  private lazy val numPendingOutstandingFutures = new AtomicInteger(0)
-  private lazy val responses = Queue.linkedNonBlocking[(Seq[S], Try[TraversableOnce[O]])]
-
-  // For testing only
-  private[executor] def outstandingFuturesQueue = outstandingFutures
+  private[executor] lazy val futureQueue = new FutureQueue[Seq[S], TraversableOnce[O]](maxWaitingFutures, maxWaitingTime, maxEmitPerExec)
 
   override def executeTick =
-    finishExecute(tick.onFailure { thr => responses.put(((Seq(), Failure(thr)))) })
+    finishExecute(Nil, tick)
 
   override def execute(state: S, data: I) =
-    finishExecute(apply(state, data).onFailure { thr => responses.put(((List(state), Failure(thr)))) })
+    finishExecute(List(state), apply(state, data))
 
-  private def finishExecute(fIn: Future[TraversableOnce[(Seq[S], Future[TraversableOnce[O]])]]) = {
-    addOutstandingFuture(handleSuccess(fIn).unit)
+  private def finishExecute(states: Seq[S], fIn: Future[TraversableOnce[(Seq[S], Future[TraversableOnce[O]])]]) = {
```

Comment:
actually, only the first `Seq[S]` is an option, so that breaks the uniformity of the typing in FutureQueue.

I wonder if `addAllFuture` is so specialized, it should stay in this class. I don't see how any user can use it generally without a big coincidence in the `FutureQueue` implementation. To me, I think so, then I think `FutureQueue` becomes very clear and simple.